### PR TITLE
[Refactoring] Add refactoring action to move type members to extension

### PR DESCRIPTION
**Pull request by ahoppen**
This adds a refactoring action to move selected members of a type to a new extension of that type. The extension is added right after the original type declaration.

Resolves [SR-6297](https://bugs.swift.org/browse/SR-6297). 
A toolchain with this refactoring action added can be downloaded [here](https://www.dropbox.com/s/4x55d58bm62jtg4/swift-LOCAL-2017-11-07-a-osx.tar.gz?dl=0).

### CODE REVIEW

File: lib/IDE/Refactoring.cpp [hunk 1]
```diff
@@ -1453,6 +1456,105 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
                                           EditConsumer).performChange();
 }
 
+/// Compute a context C such that C is a parent context of A and B.
+/// If no such context exists, return \c nullptr.
+DeclContext *getCommonParentContext(DeclContext *A, DeclContext *B) {
```
Comment (ahoppen) [inline lib/IDE/Refactoring.cpp]: Would this be interesting enough to promote it to DeclContext.h?

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

File: include/swift/Basic/SourceLoc.h [hunk 2]
```diff
@@ -96,6 +96,19 @@ class SourceRange {
...
+    SourceLoc NewStart = Start;
+    SourceLoc NewEnd = End;
+    if (Other.Start.Value.getPointer() < NewStart.Value.getPointer())
+      NewStart = Other.Start;
+    if (Other.End.Value.getPointer() > NewEnd.Value.getPointer())
+      NewEnd = Other.End;
+
+    return SourceRange(NewStart, NewEnd);
+  }
+
```
Comment (nkcsgexi) [inline include/swift/Basic/SourceLoc.h]: Can we move this outline to SourceLoc.cpp?

File: lib/IDE/Refactoring.cpp (hunk 1 above)
Comment (nkcsgexi) [reply lib/IDE/Refactoring.cpp]: That's a good idea. But similarly, please put the implementation in a .cpp file.

File: lib/IDE/Refactoring.cpp [hunk 3]
```diff
@@ -1453,6 +1456,105 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
...
+  if (A == B)
+    return A;
+
+  if (A->isChildContextOf(B))
+    return B;
+
+  if (B->isChildContextOf(A))
+    return A;
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: We don't need this `if` statement, right? the following loop covers this case.

File: lib/IDE/Refactoring.cpp [hunk 4]
```diff
@@ -1453,6 +1456,105 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
...
+    // Compute a decl context that is the parent context for all selected
+    // decls
+    DeclContext *CommonDC = nullptr;
+    for (auto DD : Info.DeclaredDecls) {
+      if (!CommonDC) {
+        CommonDC = DD.VD->getDeclContext();
+      } else {
+        CommonDC = getCommonParentContext(CommonDC, DD.VD->getDeclContext());
+      }
+    }
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: Could we move this for loop to its own function to eliminate the code duplication with the `performChange` part?

File: lib/IDE/Refactoring.cpp [hunk 5]
```diff
@@ -1453,6 +1456,105 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
...
+
+bool RefactoringActionMoveMembersToExtension::performChange() {
+  DeclContext *CommonDC = nullptr;
+  for (auto D : RangeInfo.DeclaredDecls) {
+    if (!CommonDC) {
+      CommonDC = D.VD->getDeclContext();
+    } else {
+      CommonDC = getCommonParentContext(CommonDC, D.VD->getDeclContext());
+    }
+  }
```
Comment (nkcsgexi) [inline lib/IDE/Refactoring.cpp]: I mean we can eliminate code duplication here.

File: lib/IDE/SwiftSourceDocInfo.cpp [hunk 6]
```diff
@@ -1247,7 +1261,19 @@ struct RangeResolver::Implementation {
...
+    auto NodeRange = Node.getSourceRange();
+
+    // Widen the node's source range to include all attributes to get a range
+    // match if a function with its attributes has been selected.
+    if (auto D = Node.dyn_cast<Decl *>()) {
+      for (auto Attr : D->getAttrs()) {
+        if (Attr->getRange().isValid())
+          NodeRange = NodeRange.widened(Attr->getRange());
+      }
+    }
```
Comment (nkcsgexi) [inline lib/IDE/SwiftSourceDocInfo.cpp]: Can we move this function as a member function of `Decl`? we can name it as "getRangeIncludingAttrs()".

File: lib/IDE/SwiftSourceDocInfo.cpp [hunk 7]
```diff
@@ -883,6 +884,18 @@ struct RangeResolver::Implementation {
       return llvm::any_of(StartMatches, IsCase) &&
           llvm::any_of(EndMatches, IsCase);
     }
+
+    bool isMultiDecl() {
+      if (StartMatches.empty() || EndMatches.empty())
+        return false;
+
+      // Multi-statement should have the same nominal type as a common parent
```
Comment (nkcsgexi) [inline lib/IDE/SwiftSourceDocInfo.cpp]: do you mean `Multi-decl` in the comment?

File: lib/IDE/SwiftSourceDocInfo.cpp [hunk 8]
```diff
@@ -883,6 +884,18 @@ struct RangeResolver::Implementation {
...
+      if (auto ParentDecl = Parent.dyn_cast<Decl *>())
+        if (isa<NominalTypeDecl>(ParentDecl))
+          return true;
```
Comment (nkcsgexi) [inline lib/IDE/SwiftSourceDocInfo.cpp]: `return isa<NominalTypeDecl>(ParentDecl);` is good enough here.

Comment (nkcsgexi) [reply lib/IDE/SwiftSourceDocInfo.cpp]: What happens if we select multiple top-level decls? It seems `MultiDecl` actually mean `MultiTypeMemberDecl`. If this is true, could we rename it?

**Review by nkcsgexi — COMMENTED** [review]
Looks great overall! Some minor comments inline.

**Comment by ahoppen** [comment]
Thanks for the review. I have addressed the comments. Can you have another look?

@swift-ci Please smoke test

**Comment by nkcsgexi** [comment]
Looking good! Merging.

**Pull request merged**